torch/testing/asserts.py

```diff
@@ -0,0 +1,389 @@
+import functools
+import inspect
+import sys
+import warnings
+from typing import Any, Callable, Optional, Tuple
+
+import torch
+from torch.testing._internal.common_utils import get_comparison_dtype as _get_comparison_dtype, TestCase as _TestCase
+
+__all__ = ["assert_tensors_equal", "assert_tensors_allclose"]
+
+
+# The module 'pytest' will be imported if the 'pytest' runner is used. This will only give false-positives in case
+# the test directly or indirectly imports 'pytest', but is run by another runner such as 'unittest'.
+_RUN_BY_PYTEST = "pytest" in sys.modules
+
+
+# The Usage(Error|Warning) should be raised in case the test function is not used correctly. With this the user is able
+# to differentiate between a test failure (there is a bug in the tested code) and a test error (there is a bug in the
+# test). If pytest is the test runner, we use the built-in UsageError instead our custom one.
+if _RUN_BY_PYTEST:
+    import pytest
+
+    UsageError = pytest.UsageError
+else:
+    class UsageError(Exception):
+        pass
+
+
+class UsageWarning(Warning):
+    pass
+
+
+def _get_default_rtol_and_atol(a: torch.Tensor, b: torch.Tensor) -> Tuple[float, float]:
+    try:
+        return _TestCase()._getDefaultRtolAndAtol(a.dtype, b.dtype)
+    except KeyError:
+        return 0.0, 0.0
+
+
+def _hide_internal_traceback_pytest(fn):
+    """Decorator for user-facing function that hides the internal traceback for :mod:`pytest`.
+
+    The decorator works by assigning ``__tracebackhide__ = True`` in each previous frame in case an
+    :class:`AssertionError` is encountered. If it is set manually, ``__tracebackhide__`` is not overwritten.
+
+    This is a :mod:`pytest`
+    `feature <https://docs.pytest.org/en/stable/example/simple.html#writing-well-integrated-assertion-helpers>`_
+    and thus this is a no-op if :mod:`pytest` is not present.
+    """
+    if not _RUN_BY_PYTEST:
+        return fn
+
+    @functools.wraps(fn)
+    def wrapper(*args, **kwargs):
+        try:
+            return fn(*args, **kwargs)
+        except AssertionError:
+            for frame_info in inspect.trace():
+                frame_info.frame.f_locals.setdefault("__tracebackhide__", True)
+            raise
+
+    return wrapper
+
+
+def _assert_are_tensors(a: Any, b: Any) -> None:
+    """Asserts that both inputs are tensors.
+
+    Args:
+        a (Any): First input.
+        b (Any): Second input.
+
+    Raises:
+        AssertionError: If :attr:`a` or :attr:`b` is not a :class:`~torch.Tensor`.
+    """
+    if not (isinstance(a, torch.Tensor) and isinstance(b, torch.Tensor)):
+        raise AssertionError(f"Both inputs have to be tensors, but got {type(a)} and {type(b)} instead.")
+
+
+def _assert_attributes_equal(
```

Comment:
Really nice work on this function. Great docs and docs consistency + a precise implementation.